src/conv/invokers/impl_gemm_dynamic.cpp

```diff
@@ -416,5 +416,282 @@ MakeImplGemmDynamicBackwardDataInvokerFactory<solver::TunableImplicitGemmGTCDyna
+
+    bool need_set_zero = false;
+    if(y < stride_h || x < stride_w || dilation_h != 1 || dilation_w != 1)
+        need_set_zero = true;
+    need_set_zero |= config.gemm_k_global_split > 0;
+
+    std::vector<OpKernelArg> opShapeArgs;
+    opShapeArgs.emplace_back(hi);
+    opShapeArgs.emplace_back(wi);
+    opShapeArgs.emplace_back(n);
+    opShapeArgs.emplace_back(k / group);
+    opShapeArgs.emplace_back(c / group);
+    opShapeArgs.emplace_back(ho);
+    opShapeArgs.emplace_back(wo);
+    opShapeArgs.emplace_back(stride_h);
+    opShapeArgs.emplace_back(stride_w);
+    opShapeArgs.emplace_back(dilation_h);
+    opShapeArgs.emplace_back(dilation_w);
+    opShapeArgs.emplace_back(pad_h);
+    opShapeArgs.emplace_back(pad_w);
+    opShapeArgs.emplace_back(y);
+    opShapeArgs.emplace_back(x);
+
+    opShapeArgs.emplace_back(dtile_iy);
+    opShapeArgs.emplace_back(dtile_ix);
+    opShapeArgs.emplace_back(dilation_h / gcd_stride_dilation_h);
+    opShapeArgs.emplace_back(dilation_w / gcd_stride_dilation_w);
+    opShapeArgs.emplace_back(y_tilda);
+    opShapeArgs.emplace_back(x_tilda);
+    opShapeArgs.emplace_back(dtile_h);
+    opShapeArgs.emplace_back(dtile_w);
+    opShapeArgs.emplace_back(dslice_y);
+    opShapeArgs.emplace_back(dslice_x);
+
+    opShapeArgs.emplace_back(h_tilda_slice);
+    opShapeArgs.emplace_back(w_tilda_slice);
+    opShapeArgs.emplace_back(h_tilda_left);
+    opShapeArgs.emplace_back(w_tilda_left);
+    opShapeArgs.emplace_back(group);
+
+    opShapeArgs.emplace_back(mdiv_0.magic);
+    opShapeArgs.emplace_back(mdiv_1.magic);
+    opShapeArgs.emplace_back(mdiv_2.magic);
+    opShapeArgs.emplace_back(mdiv_3.magic);
+    opShapeArgs.emplace_back(shift_pack_0);
+    opShapeArgs.emplace_back(config.gemm_k_global_split);
+
+    return [opShapeArgs, need_set_zero](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle, const AnyInvokeParams& primitive_parameters) {
+            decltype(auto) data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();
+            const auto& tensors     = data_ctx.tensors;
+            const auto ker          = handle.Run(kernels[0]);
+            float elapsed           = 0;
+
+            std::vector<OpKernelArg> opArgs;
+            opArgs.reserve(3 + opShapeArgs.size()); // Avoids vector resize.
+            opArgs.emplace_back(tensors.out);
+            opArgs.emplace_back(tensors.w);
+            opArgs.emplace_back(tensors.in);
+
+            if(need_set_zero)
+            {
+                float zero = 0.f;
+                SetTensor(handle, tensors.outDesc, tensors.out, &zero);
+                if(handle.IsProfilingEnabled())
+                    elapsed += handle.GetKernelTime();
+            }
+
+            std::transform(opShapeArgs.begin(),
+                           opShapeArgs.end(),
+                           std::back_inserter(opArgs),
+                           [](const OpKernelArg& arg) { return arg; });
```

Comment:
It seems like we are copying the whole contents of `opShapeArgs` to `opArgs` in the Invoker. Can we prepare the whole `opArg` in the Factory, pass it by-copy to the Invoker and just fill in three pointers here?

Reply:
OK. I changed the code in both `MakeImplGemmDynamicForwardXdlopsNHWCInvokerFactory` `MakeImplGemmDynamicBackwardDataXdlopsNHWCInvokerFactory`, only use `opArgs` out side the lambda and change corresponding pointer in it.

Reply:
[Future][Optimization] What about changing all other Asm iGemm invokers like this? I can open a ticket.

Reply:
[Future][Optimization] @carlushuang What about changing all other Asm iGemm invokers like this? I can open a ticket.

Reply:
@atamazov OK

Reply:
@carlushuang Should I open a ticket or you'll do that yourself? :wink:

Reply:
@atamazov sure

Reply:
@carlushuang That would be fine, please go ahead!

Reply:
#1005